chainercv/datasets/siamese_dataset.py

```diff
@@ -0,0 +1,130 @@
+    for i, label in enumerate(labels):
+        d[label].append(i)
+    return d
+
+
+class SiameseDataset(chainer.dataset.DatasetMixin):
+
+    """A dataset that returns samples fetched from two datasets.
+
+    The dataset returns samples from the two base datasets.
+    If :obj:`pos_ratio` is not :obj:`None`,
+    :class:`SiameseDataset` can be configured to return positive
+    pairs at the ratio of :obj:`pos_ratio` and negative pairs at the ratio
+    of :obj:`1 - pos_ratio`.
+    In this mode, the base datasets are assumed to be label datasets that
+    return an image and a label as a sample.
+
+    Args:
+        dataset_0: The first base dataset.
+        dataset_1: The second base dataset.
+        pos_ratio (float): If this is not :obj:`None`,
+            this dataset tries to construct positive pairs at the
+            given rate. If :obj:`None`,
+            this dataset randomly samples examples from the base
+            datasets. The default value is :obj:`None`.
+        length (int): The length of this dataset. If :obj:`None`,
+            the length of the first base dataset is the length of this
+            dataset.
+        labels_0 (numpy.ndarray): The labels associated to
+            the first base dataset. The length should be the same as
+            the length of the first dataset. If this is :obj:`None`,
+            the labels are automatically fetched using the following
+            line of code: :obj:`[ex[1] for ex in dataset_0]`.
+            By setting :obj:`labels_0` and skipping the fetching
+            iteration, the computation cost can be reduced.
+            Also, if :obj:`pos_ratio` is :obj:`None`, this value
+            is ignored. The default value is :obj:`None`.
+            If :obj:`labels_1` is spcified and
+            :obj:`dataset_0` and :obj:`dataset_1` are the same,
+            :obj:`labels_0` can be skipped.
+        labels_1 (numpy.ndarray): The labels associated to
+            the second base dataset. If :obj:`labels_0` is spcified and
+            :obj:`dataset_0` and :obj:`dataset_1` are the same,
+            :obj:`labels_1` can be skipped.
+            Please consult the explanation for :obj:`labels_0`.
+
+    """
+
+    def __init__(self, dataset_0, dataset_1,
+                 pos_ratio=None, length=None, labels_0=None, labels_1=None):
+        self._dataset_0 = dataset_0
+        self._dataset_1 = dataset_1
+        self._pos_ratio = pos_ratio
+        if length is None:
+            length = len(self._dataset_0)
+        self._length = length
+
+        if pos_ratio is not None:
+            if labels_1 is None and dataset_0 is dataset_1:
+                labels_1 = labels_0
+            if labels_0 is None and dataset_0 is dataset_1:
+                labels_0 = labels_1
+            if labels_0 is None:
+                labels_0 = [example[1] for example in dataset_0]
+            if labels_1 is None:
+                labels_1 = [example[1] for example in dataset_1]
```

Comment:
In the case of `labels_0 is None`, `labels_1 is None` and `dataset_0 is dataset_1`,
```python
        if pos_ratio is not None:
            if labels_1 is None and dataset_0 is dataset_1:
                labels_1 = labels_0  # labels_1 = None
            if labels_0 is None and dataset_0 is dataset_1:
                labels_0 = labels_1  # labels_0 = None
            if labels_0 is None:
                labels_0 = [example[1] for example in dataset_0]  # fetch labels from dataset_0 <- OK
            if labels_1 is None:
                labels_1 = [example[1] for example in dataset_1]  # fetch labels from dataset_1 <- redundant
```